pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -71,18 +78,30 @@ public SelectionOperatorService(QueryContext queryContext, DataSchema dataSchema
     _offset = queryContext.getOffset();
     _numRowsToKeep = _offset + queryContext.getLimit();
     assert queryContext.getOrderByExpressions() != null;
+    _comparator = OrderByComparatorFactory.getComparator(queryContext.getOrderByExpressions(),
+        _queryContext.isNullHandlingEnabled());
     _rows = new PriorityQueue<>(Math.min(_numRowsToKeep, SelectionOperatorUtils.MAX_ROW_HOLDER_INITIAL_CAPACITY),
         OrderByComparatorFactory.getComparator(queryContext.getOrderByExpressions(),
             _queryContext.isNullHandlingEnabled()).reversed());
+
+    _sortedRows = new ArrayList<>();
   }
 
   /**
    * Reduces a collection of {@link DataTable}s to selection rows for selection queries with <code>ORDER BY</code>.
-   * TODO: Do merge sort after releasing 0.13.0 when server side results are sorted
-   *       Can also consider adding a data table metadata to indicate whether the server side results are sorted
+   * Do merge sort when server side results are sorted
    */
   public void reduceWithOrdering(Collection<DataTable> dataTables) {
     for (DataTable dataTable : dataTables) {
```

Comment:
Consider first looping over all data tables and see if all of them are sorted, then we don't need to handle the case where half data tables are sorted while others unsorted